src/core/installAddon.js

```diff
@@ -0,0 +1,192 @@
+/* global CustomEvent, document, window */
+import { connect } from 'react-redux';
+import config from 'config';
+
+import log from 'core/logger';
+import tracking, { getAction } from 'core/tracking';
+import {
+  CLOSE_INFO,
+  DISABLED,
+  DOWNLOAD_FAILED,
+  DOWNLOAD_PROGRESS,
+  ENABLED,
+  ERROR,
+  FATAL_ERROR,
+  FATAL_INSTALL_ERROR,
+  FATAL_UNINSTALL_ERROR,
+  INSTALL_CATEGORY,
+  INSTALL_ERROR,
+  INSTALL_FAILED,
+  INSTALL_STATE,
+  SET_ENABLE_NOT_AVAILABLE,
+  SHOW_INFO,
+  START_DOWNLOAD,
+  UNINSTALL_CATEGORY,
+  UNINSTALLED,
+  UNINSTALLING,
+} from 'core/constants';
+import * as addonManager from 'core/addonManager';
+
+export function makeProgressHandler(dispatch, guid) {
+  return (addonInstall, e) => {
+    if (addonInstall.state === 'STATE_DOWNLOADING') {
+      const downloadProgress = parseInt(
+        (100 * addonInstall.progress) / addonInstall.maxProgress, 10);
+      dispatch({ type: DOWNLOAD_PROGRESS, payload: { guid, downloadProgress } });
+    } else if (e.type === 'onDownloadFailed') {
+      dispatch({ type: INSTALL_ERROR, payload: { guid, error: DOWNLOAD_FAILED } });
+    } else if (e.type === 'onInstallFailed') {
+      dispatch({ type: INSTALL_ERROR, payload: { guid, error: INSTALL_FAILED } });
+    }
+  };
+}
+
+export function makeMapDispatchToProps({ src }) {
+  return function mapDispatchToProps(
+    dispatch,
+    {
+      _addonManager = addonManager,
+      _config = config,
+      _dispatchEvent,
+      _tracking = tracking,
+      ...ownProps
+    } = {},
+  ) {
+    if (config.get('server')) {
+      return {};
+    }
+
+    // Set the default here otherwise server code will blow up.
+    const dispatchEvent = _dispatchEvent || document.dispatchEvent;
+
+    function showInfo({ name, iconUrl, i18n }) {
+      if (_config.has('useUiTour') && _config.get('useUiTour')) {
+        dispatchEvent(new CustomEvent('mozUITour', {
+          bubbles: true,
+          detail: {
+            action: 'showInfo',
+            data: {
+              target: 'appMenu',
+              icon: iconUrl,
+              title: i18n.gettext('Your add-on is ready'),
+              text: i18n.sprintf(
+                i18n.gettext('Now you can access %(name)s from the toolbar.'),
+              { name }),
+              buttons: [{ label: i18n.gettext('OK!'), callbackID: 'add-on-installed' }],
+            },
+          },
+        }));
+      } else {
+        dispatch({
+          type: SHOW_INFO,
+          payload: {
+            addonName: name,
+            imageURL: iconUrl,
+            closeAction: () => {
+              dispatch({ type: CLOSE_INFO });
+            },
+          },
+        });
+      }
+    }
+
+    return {
+      setCurrentStatus() {
+        const { guid, installURL } = ownProps;
+        const payload = { guid, url: installURL };
+        return _addonManager.getAddon(guid)
+        .then(
+          (addon) => {
```

Comment:
Sure, you end up with a `}, () => {` for the catch part but that's how I normally would have written it.